sql/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java

```diff
@@ -167,10 +176,16 @@ public boolean moveNext() {
                 if (line == null) {
                     closeCurrentReader();
                     return moveNext();
-                } else {
-                    lineContext.rawSource(line.getBytes(StandardCharsets.UTF_8));
+                }
+
+                if (csvHeader.isPresent()) {
+                    processAsCSV(csvHeader.get(), line);
                     return true;
+                } else {
+                    processAsJsonLine(line);
```

Comment:
Would be great if we could encapsulate this functionality in the LineContext implementation for csv/json. See above.